### ISSUE TITLE
Python 3.10: TypeError: __add__() got an unexpected keyword argument 'context'

### ISSUE THREAD

**Issue by rndstr** [eprbell/rp2#25]
Running the example i'm getting this error (tested with Python3.9 and Python3.10)

```
 $ rp2_us -m fifo -o output -p crypto_example_ crypto_example.config crypto_example.ods
INFO: Country: us
INFO: Accounting Method: fifo
INFO: Configuration file: crypto_example.config
INFO: Input file: crypto_example.ods
INFO: Processing BTC
ERROR: Fatal exception occurred:
Traceback (most recent call last):
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/rp2_main.py", line 93, in _rp2_main_internal
    input_data: InputData = parse_ods(configuration=configuration, asset=asset, input_file_handle=input_file_handle)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/ods_parser.py", line 128, in parse_ods
    _create_and_process_transaction(
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/ods_parser.py", line 168, in _create_and_process_transaction
    transaction: AbstractTransaction = _create_transaction(configuration, current_table_type, internal_id, row_values)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/ods_parser.py", line 277, in _create_transaction
    transaction = InTransaction(**argument_pack)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/in_transaction.py", line 52, in __init__
    super().__init__(configuration, timestamp, asset, transaction_type, spot_price, internal_id, unique_id, notes)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/abstract_transaction.py", line 41, in __init__
    self.__spot_price: RP2Decimal = configuration.type_check_positive_decimal("spot_price", spot_price)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/configuration.py", line 276, in type_check_positive_decimal
    if result < ZERO:
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/rp2_decimal.py", line 59, in __lt__
    return not self.__ge__(other)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/rp2_decimal.py", line 48, in __ge__
    return (self - other).quantize(CRYPTO_DECIMAL_MASK).__ge__(ZERO)
  File "/home/rndstr/venv/lib/python3.9/site-packages/rp2/rp2_decimal.py", line 69, in __sub__
    return RP2Decimal(Decimal.__sub__(self, other))
  File "/usr/lib/python3.9/_pydecimal.py", line 1257, in __sub__
    return self.__add__(other.copy_negate(), context=context)
TypeError: __add__() got an unexpected keyword argument 'context'
INFO: Log file: ./log/rp2_2022_04_10_15_52_55_045185.log
INFO: Generated output directory: output
INFO: Done
```

**Comment by eprbell** [comment]
Which version of RP2 are you using? If not the latest, could you upgrade and try again? Also make sure the example ODS and config are the latest version.

**Comment by rndstr** [comment]
I set everything up yesterday and was using 0.9.29 installed with `pip`.

**Comment by rndstr** [comment]
It looks like you are [overriding Decimal.\_\_add__()](https://github.com/eprbell/rp2/blob/67b00f9983c5e9f2baba122533fa369d454b1dae/src/rp2/rp2_decimal.py#L61) but do not provide the `context` param which is in the `Decimal` class in Python3.9/3.10:

https://github.com/python/cpython/blob/89697f7374ea947ebe8e36131e2d3e21fff6fa1d/Lib/_pydecimal.py#L1157

and then being called virtually

https://github.com/python/cpython/blob/f33e2c87a83917b5139d97fd8ef7cba7223ebef5/Lib/_pydecimal.py#L1257

not sure why it would work for others--if that's the case--though.

**Comment by eprbell** [comment]
Interesting: haven't seen that before. The crypto_example.ods test is ran as part of continuous integration at every checkin on Python 3.7, 3.8, 3.9 and on Linux, Windows and Mac. A few more questions:
- what OS and OS version are you using?
- did you install Python as described in the docs or did you use a Python that came with the OS?

**Comment by eprbell** [comment]
The unit tests are failing in the PR:
```
Traceback (most recent call last):
  File "/home/runner/work/rp2/rp2/src/rp2/rp2_main.py", line 93, in _rp2_main_internal
    input_data: InputData = parse_ods(configuration=configuration, asset=asset, input_file_handle=input_file_handle)
  File "/home/runner/work/rp2/rp2/src/rp2/ods_parser.py", line 128, in parse_ods
    _create_and_process_transaction(
  File "/home/runner/work/rp2/rp2/src/rp2/ods_parser.py", line 168, in _create_and_process_transaction
    transaction: AbstractTransaction = _create_transaction(configuration, current_table_type, internal_id, row_values)
  File "/home/runner/work/rp2/rp2/src/rp2/ods_parser.py", line 277, in _create_transaction
    transaction = InTransaction(**argument_pack)
  File "/home/runner/work/rp2/rp2/src/rp2/in_transaction.py", line 81, in __init__
    self.__fiat_in_with_fee = self.__fiat_in_no_fee + self.__fiat_fee
  File "/home/runner/work/rp2/rp2/src/rp2/rp2_decimal.py", line 64, in __add__
    return RP2Decimal(Decimal.__add__(self, other, context))
TypeError: expected 1 argument, got 2
```
The context is probably something that got added in Python 3.10 (unit tests are running on 3.7, 3.8 and 3.9).

**Comment by rndstr** [comment]
Odd, I see it in 3.7 as well https://github.com/python/cpython/blob/3.7/Lib/_pydecimal.py#L1157

**Comment by rndstr** [comment]
Is there some sort of non-strict mode in Python that would allow extra arguments in this particular instance even though it's not part of the signature?

**Comment by eprbell** [comment]
Not sure... but it looks like RP2 is not working yet on Python 3.10. If you find a way to make the PR work in a backward compatible way, I'll accept it. For now the (not ideal) workaround is to use Python 3.9.

**Comment by eprbell** [comment]
I tried with Python 3.10 on my personal machine and I couldn't repro this problem. I also added Python 3.10 to continuous integration on Windows, Mac and Linux and it all works well (see https://github.com/eprbell/rp2/commit/0d7b01a7649d8f50c65c28c85e323e92ad5ce2d5). I think the issue you are seeing might be due to some specific problem with your Python installation.

**Comment by eprbell** [comment]
Here is continuous integration running on Python 3.10:
* Windows: https://github.com/eprbell/rp2/runs/5993615909?check_suite_focus=true
* Mac: https://github.com/eprbell/rp2/runs/5993616027?check_suite_focus=true
* Linux: https://github.com/eprbell/rp2/runs/5993615670?check_suite_focus=true

You may want to try uninstalling python and reinstalling it.

**Comment by eprbell** [comment]
I don't think there is anything else actionable on this issue. OK to close?

**Comment by antonok-edm** [comment]
FWIW, I think it has to due with the fact that there are two possible backends for the decimal module, see [cpython/Lib/decimal.py](https://github.com/python/cpython/blob/d5ebf8b71fd18d7a1f2f6b670a2c18749dc2b55e/Lib/decimal.py).

I confirmed that on Arch Linux it's loading from `_pydecimal`. I assume that elsewhere it's loading from `_decimal` instead.

**Comment by antonok-edm** [comment]
Found it!

```
> python
Python 3.11.8 (main, Feb 12 2024, 14:50:05) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from _decimal import *
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: libmpdec.so.4: cannot open shared object file: No such file or directory
```
```
> pacman -Ss mpdec
extra/mpdecimal 4.0.0-2
    Package for correctly-rounded arbitrary precision decimal floating point
    arithmetic
> sudo pacman -Syu mpdecimal
...
> python
Python 3.11.8 (main, Feb 12 2024, 14:50:05) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from _decimal import *
>>>
```
...and `rp2` appears to be working after this process as well.

TLDR: `mpdecimal` is a required dependency of `rp2`. I guess on Ubuntu it's either installed by default with the OS or with the distribution of Python. On Arch it must be installed separately.

**Comment by eprbell** [comment]
Fantastic! Thanks for the good detective work :-)

**Comment by antonok-edm** [comment]
I filed [a bug](https://github.com/python/cpython/issues/117056) against Python; maybe this could be fixed upstream.

**Comment by eprbell** [comment]
Good idea. If you also want to submit a PR to add a dependency for RP2 I'll approve it.

**Comment by eprbell** [comment]
This should now be resolved (see #108).
